### PR TITLE
Multi-objective bugfixes and parallel execution

### PR DESCRIPTION
**Pull request by cengique**
Hi,

I've been working on GODLIKE to get it to work with my multi-objective functions and enable parallel processing. So in summary these are my changes:

* Some bugfixes when using a single function that returns multiple objectives. Better multi-objective function detection.
* Added a new option "UseParallel", which uses Matlab's parfor command to run function evaluations in parallel. Saves a lot of time on multi-threaded or cluster computers.

Would you like to merge these back on your master?

Thanks for your work, it's quite well written except some tiny bugs I found and hopefully fixed correctly. I used your demo as a test case, and it still functions correctly.
-Cengiz

### CODE REVIEW

**Comment by ambrix** [comment]
I have to make a multi-objective optimization . I have a function called statement, my_obj , where there are 2 decision variables and 3 objective functions . Now I call this function by another m-file :

[ s0 , FVAL ] = GODLIKE ( @my_obj , PS , lb , ub , ' ASA ' , options) ;

but I always get this error :

??? Error using == > cellfun
Non - scalar in Uniform output , at index 1 , output 1 .
Set ' UniformOutput ' to false .

How should I give the 3 objective functions to algorithm GODLIKE ? In this moment I in the function statement, my_obj I have had write:

of1 = 1 -E ; % or of(1)= ...
of2 = C ;
of3 = D ;

where E , ​​C , D are dependent on the 2 decision variables .
Can you help me? Thanks

**Comment by cengique** [comment]
Hi @ambrix, if I'm not mistaken, you need to return the 3 objective values in a single row array. Here's an example:

```matlab
function objs = myobj
...
objs = [ 1 - E, C, D]
```
In your example, the commented-out version, the syntax `of(1)=...` would have worked as well. Let me know if this helps.

**Comment by ambrix** [comment]
Hi @cengique, unfortunately it does not work in everyone of the two ways (objs = [1-E, C, D] & objs (1) = ...). The problem is how to pass these parameters and it seems that GODLIKE want it built a CELL-ARRAY, but I am trying in every way but I can not work it! :(

**Comment by cengique** [comment]
@ambrix : please open a new **issue** in my fork and post your `myobj` function there. 

You don't need to return a cell array to make this work. If your objective functions are completely separate, then you can provide separate objective functions in a cell array (e.g., `{@myobj1, @myobj2}`. But in my fork, I have fixed it such that one function can return multiple objectives. You can also try to set the `NumObjectives` option to specify how many objectives you're returning.

**Comment by ambrix** [comment]
@cengique: excuse me but i am new in github and i don't know how to create a fork!

I have tried to manually enter numObjective but the algorithm is still not working. Even inserting of objfun = {@ myobj1, ...} or objfun = [of1,of2,of3], GODLIKE not working.
I want to point out that for a single objective function (of1 or of2 or of3 only), the algorithm works well. Therefore, it is only a problem of the multiobjective! Can you help me? Maybe I can send you the m-files via email or I don't know how I can fix, I'm trying all ways!!

**Comment by cengique** [comment]
@ambrix : To create an issue, please go to https://github.com/cengique/FEX-GODLIKE and click on "Issues" on the right-side menu. Then click on the "New Issue" button and write a brief description and attach your objective function as a file. Thanks!

**Comment by rodyo** [comment]
Hi! 

Again, sorry for the incredibly late reply. 

It seems you've found & fixed copious bugs that I'm more than pleased to merge into master, thanks a lot for that!  

Do you have any experience with how that works on the Mathworks File Exchange? I mean, I want to give you some credit eventually, both here and on FEX. On github it's practically automatic, but on the FEX...

So, I should have looked at this before my recent refactoring efforts :)  Merging is now a bit more involved...oh well, I'll get to it. 

My main intentions with GODLIKE are as follows: 
- refactor codebase so that adding new algorithms and changing existing ones becomes possible, and preferably also easy to do (currently DE, PSO, GA and ASA are sort of hardcoded everywhere)
- Implement support for (non)linear constraints

After this merge, would you like to collaborate on that?

File: GODLIKE.m [hunk 1]
```diff
@@ -989,17 +989,19 @@ function check_input(varargin)
...
-                            'is %d, upper bound is %d.\n'], popsize, options.GODLIKE.ItersLb, ...
-                            options.GODLIKE.ItersUb);
+                           ['Performing single-objective optimization.\n']);
                     elseif multi
                         fprintf(1,...                            
-                           ['Performing multi-objective optimization, with %d objectives.\n',...
-                            'Total population size is %d individuals. Lower bounds on\nalgorithm ',...
-                            'iterations is %d, upper bound is %d.\n'], options.num_objectives,...
-                            popsize, options.GODLIKE.ItersLb, options.GODLIKE.ItersUb);
+                           ['Performing multi-objective optimization, with %d objectives.\n']);
```
Comment (rodyo) [inline GODLIKE.m]: you forgot the options.num_objectives there...

File: pop_single.m [hunk 2]
```diff
@@ -706,52 +709,61 @@ function replace_parents(pop)
...
+
+            pop_inputs = ...
+              pop.pop_data.offspring_population(sites, :);
+
             % evaluate all functions for each population member in
             % parallel or in serial
-            fvs = pop.pop_data.offspring_population(sites, :);
             if pop.options.UseParallel
               parfor pop_num=1:num_pop
-                evaluate_one_function(pop, sites, pop_num);
```
Comment (rodyo) [inline pop_single.m]: I have a feeling this way of parallelizing is inefficient...wouldn't it be better to evaluate the different algorithms in parallel, in their entirety? e.g., DE on 1 core, ASA on another, GA on the 3rd, etc. Feels like there should be less overhead that way, but profiling is required

File: GODLIKE.m [hunk 3]
```diff
@@ -54,7 +54,9 @@
 %
 %   popsize     positive integer. Indicates the TOTAL population 
 %               size, that is, the number of individuals of all 
-%               populations combined.
```
Comment (rodyo) [inline GODLIKE.m]: I'm not so sure about this feature; let me think a bit more on it. 

Do you have some use cases/examples where this actually improves the performance/reliability?

**Comment by rodyo** [comment]
Merging was too much effort -- went by all your changes one by one, and copied to the correct places.

File: GODLIKE.m (hunk 3 above)
Comment (cengique) [inline GODLIKE.m]: Hi @rodyo,

I assume this one is about passing an array of popsize for each algorithm. This is useful for the parallel execution feature. I was running on a 64-core machine and wanted to make sure that each algorithm's population size is a multiple of 64. By default, GODLIKE would randomly pick population sizes for each algorithm totaling to popsize. But in that case, 64 cores would be used unevenly and some of them would have to wait for others to finish. With this feature, the user can specify how many to run in each algorithm. Another solution would be to have a new option to ask to divide population equally among the algorithms, but I found that this solution is more flexible and doesn't require adding a new option.

**Comment by cengique** [comment]
Thank you for putting in the work to merge my changes and not ignoring them! Greatly appreciated! I'll respond to your email too.

**Comment by cengique** [comment]
@rodyo Oh your message was here, not an email. :) Yeah, I would appreciate if you list me as a contributor on File Exchange. And I am still using GODLIKE actively while finishing a project, but I'm not I can commit to major rework of the code. Still, I would be happy to support your work by testing, etc.

By the way, I see you took my changes, but our branches still look separate. So, I'll try to make a fake merge commit so they are unified. Are yoy done merging my changes?

**Comment by rodyo** [comment]
Yes, except for the vector-valued popsizes bit...I'll put in those changes as well, I think it's a good idea, if only to give people that choice :)

**Comment by cengique** [comment]
Ok, let me know when you patch it so I can record the merge between our branches.

**Comment by rodyo** [comment]
Done!

**Pull request closed without merging**